test/cpp/naming/resolver_component_tests_runner_invoker_common.cc

```diff
@@ -0,0 +1,137 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <grpc/grpc.h>
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+#include <grpc/support/string_util.h>
+#include <signal.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <gflags/gflags.h>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include "test/cpp/util/subprocess.h"
+#include "test/cpp/util/test_config.h"
+
+#include "test/cpp/naming/resolver_component_tests_runner_invoker_common.h"
+
+extern "C" {
+#include "test/core/util/port.h"
+}
+
+using grpc::SubProcess;
+
+static volatile sig_atomic_t abort_wait_for_child = 0;
+
+static void sighandler(int sig) { abort_wait_for_child = 1; }
+
+static void register_sighandler() {
+  struct sigaction act;
+  memset(&act, 0, sizeof(act));
+  act.sa_handler = sighandler;
+  sigaction(SIGINT, &act, NULL);
+  sigaction(SIGTERM, &act, NULL);
+}
+
+namespace {
+
+const int kTestTimeoutSeconds = 60 * 2;
+
+void RunSigHandlingThread(SubProcess *test_driver, gpr_mu *test_driver_mu,
+                          gpr_cv *test_driver_cv, int *test_driver_done) {
+  for (size_t i = 0; i < kTestTimeoutSeconds && !abort_wait_for_child; i++) {
+    gpr_mu_lock(test_driver_mu);
+    if (*test_driver_done) {
+      gpr_mu_unlock(test_driver_mu);
+      return;
+    }
+    gpr_cv_wait(test_driver_cv, test_driver_mu,
+                gpr_time_add(gpr_now(GPR_CLOCK_REALTIME),
+                             gpr_time_from_seconds(1, GPR_TIMESPAN)));
+    gpr_mu_unlock(test_driver_mu);
+  }
+  gpr_log(GPR_DEBUG,
+          "Test timeout reached or received signal. Interrupting test driver "
+          "child process.");
+  test_driver->Interrupt();
+  return;
+}
+}
+
+namespace grpc {
+
+namespace testing {
+
+void InvokeResolverComponentTestsRunner(std::string test_runner_bin_path,
+                                        std::string test_bin_path,
+                                        std::string dns_server_bin_path,
+                                        std::string records_config_path) {
+  grpc_init();
+  int test_dns_server_port = grpc_pick_unused_port_or_die();
+
+  SubProcess *test_driver = new SubProcess(
+      {test_runner_bin_path, "--test_bin_path=" + test_bin_path,
+       "--dns_server_bin_path=" + dns_server_bin_path,
+       "--records_config_path=" + records_config_path,
+       "--test_dns_server_port=" + std::to_string(test_dns_server_port)});
+  gpr_mu test_driver_mu;
+  gpr_mu_init(&test_driver_mu);
+  gpr_cv test_driver_cv;
+  gpr_cv_init(&test_driver_cv);
+  int test_driver_done = 0;
+  register_sighandler();
+  std::thread sig_handling_thread(RunSigHandlingThread, test_driver,
```

Comment:
Is it actually necessary to spawn a separate thread for the signal handler?  Why not do it in the main thread?  I'm thinking that every time through the loop, we can wait for a second and then check whether the test has finished.

Reply:
The only problem I can see is the main thread can't periodically poll the "interrupted" flag since it's blocked indefinitely in it's `Join` call (no other way to check if the test is done that I can see except for checking a `SIG_CHILD`), and it also can't break out from it's `Join` call after a minute or two has passed. I wanted to use the background thread originally because it allows forwarding `SIGTERM`/`SIGINT` to the child process after only a short delay, setting an overall timeout in the test, and allowing the  test to terminate quickly if is succeeds in << 1 second. I also wasn't sure about calling `Subprocess->Interrupt()` in the signal handler, just to be safe, since I wanted to avoid accidentally locking on something or causing a race. Does this sound ok?

Reply:
Ah, I didn't realize that there wasn't a form of `SubProcess::Join()` that takes a timeout.  In that case, I guess this is fine as-is.  It's a shame that we need so much machinery to do something so simple, but I don't see an obviously better approach.